<a href="https://colab.research.google.com/github/remun5/Fashion-MNIST-Keras-LeNET-CNN/blob/master/Fashion_MNIST_LeNet_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Fashion MNIST Dataset with Convolutional Neural Network:**

In this project, we will again use the Fashion MNIST dataset and implement the LeNet-5 Convolutional Neural Network based learning. LetNet-5 architecture can found in the paper "Gradient-Based Learning Applied to Document. Recognition. Yann LeCun, Léeon Bottou, Yoshua Bengio, and Patrick Haffner". The idea of this was to process handwriting. However, since Fashion MNIST dataset is also a low resolution MNIST like dataset, we will try to implement it and compare it to the previously implemented simple neural network.

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from keras.models import Model
from keras.datasets import fashion_mnist
from keras.utils import np_utils
from keras.layers import Input, Conv2D, AveragePooling2D, Flatten, Dense

In [ ]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

In [ ]:
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1)
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1)

In [ ]:
y_train_enc = np_utils.to_categorical(y_train)
y_test_enc = np_utils.to_categorical(y_test)

### **LeNet 5:**

In [ ]:
def LeNetCNN(input_shape):
  X_input = Input(shape=input_shape)
  X = Conv2D(filters=6, kernel_size=(5,5), activation='tanh')(X_input)
  X = AveragePooling2D(pool_size = (2,2), strides=(2,2))(X)
  X = Conv2D(filters=16, kernel_size=(5,5), strides=(1,1), activation='tanh')(X)
  X = AveragePooling2D(pool_size=(2,2), strides=(2,2))(X)
  X = Flatten()(X)
  X = Dense(units=120, activation='tanh')(X)
  X = Dense(units=84, activation='tanh')(X)
  X = Dense(units=10, activation='softmax')(X)
  model = Model(inputs = X_input, outputs = X, name='LeNet')
  return model

In [ ]:
mnist_LeNet = LeNetCNN((28, 28, 1))

In [ ]:
print(mnist_LeNet.summary())

Model: "LeNet"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_12 (InputLayer)        [(None, 28, 28, 1)]       0         
_________________________________________________________________
conv2d_14 (Conv2D)           (None, 24, 24, 6)         156       
_________________________________________________________________
average_pooling2d_6 (Average (None, 12, 12, 6)         0         
_________________________________________________________________
conv2d_15 (Conv2D)           (None, 8, 8, 16)          2416      
_________________________________________________________________
average_pooling2d_7 (Average (None, 4, 4, 16)          0         
_________________________________________________________________
flatten_4 (Flatten)          (None, 256)               0         
_________________________________________________________________
dense_7 (Dense)              (None, 120)               30840 

In [ ]:
mnist_LeNet.compile(optimizer = "Adam", loss = "categorical_crossentropy", metrics = ["accuracy"])

In [ ]:
mnist_LeNet.fit(x = X_train, y = y_train_enc, epochs = 10, batch_size = 32)

Epoch 1/10
1875/1875 [==============================] - 27s 14ms/step - loss: 0.2474 - accuracy: 0.9071
Epoch 2/10
1875/1875 [==============================] - 27s 15ms/step - loss: 0.2353 - accuracy: 0.9119
Epoch 3/10
1875/1875 [==============================] - 27s 14ms/step - loss: 0.2335 - accuracy: 0.9121
Epoch 4/10
1875/1875 [==============================] - 27s 15ms/step - loss: 0.2197 - accuracy: 0.9176
Epoch 5/10
1875/1875 [==============================] - 27s 15ms/step - loss: 0.2138 - accuracy: 0.9199
Epoch 6/10
1875/1875 [==============================] - 27s 14ms/step - loss: 0.2078 - accuracy: 0.9216
Epoch 7/10
1875/1875 [==============================] - 27s 15ms/step - loss: 0.2019 - accuracy: 0.9239
Epoch 8/10
1875/1875 [==============================] - 27s 15ms/step - loss: 0.1957 - accuracy: 0.9271
Epoch 9/10
1875/1875 [==============================] - 27s 15ms/step - loss: 0.1912 - accuracy: 0.9279
Epoch 10/10
1875/1875 [==============================] - 27s 15m

In [ ]:
preds_train = mnist_LeNet.evaluate(x=X_train, y = y_train_enc)
print('Training::')
print ("Loss = " + str(preds_train[0]))
print ("Training Accuracy = " + str(preds_train[1]))

preds = mnist_LeNet.evaluate(x = X_test, y = y_test_enc)
print('Testing::')
print ("Loss = " + str(preds[0]))
print ("Test Accuracy = " + str(preds[1]))

1875/1875 [==============================] - 14s 7ms/step - loss: 0.1849 - accuracy: 0.9305
Training::
Loss = 0.18494786322116852
Training Accuracy = 0.9305499792098999
313/313 [==============================] - 2s 8ms/step - loss: 0.3070 - accuracy: 0.8939
Testing::
Loss = 0.30698275566101074
Test Accuracy = 0.8938999772071838


Here, as you can see the training and testing accuracies have improved from 91% training and 88% test accracies to 93% training and 89% test accuracies.